### PR TITLE
Improve revive calculation

### PR DESCRIPTION
**Pull request by gabrielhartmann**
When doing scale tests I noticed occasional revives where I expected none.  The fact that we were getting the work candidates, then processing offers then deciding whether we needed to revive was causing a situation where the Steps changed status.  So...
* Move revive calculation right next to candidate selection

There was also the fact that things move in and out of the candidate set, but shouldn't have revives called on them.  Once something moves into the active workset it should only leave upon completion.  So...
* We merge the steps which are in progress (not complete and not pending) into the active work set

N.B. All tasks will be detected as PENDING and in need of revive as part of the getCandidates call.

screen shot 2017-11-01 at 4 04 10 pm

Scale testing confirms no unexpected revives.  20 schedulers with 80 tasks each crash looping have only a single revive (when they start and go from no work to some work).

### CODE REVIEW

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/ReviveManager.java [hunk 1]
```diff
@@ -64,17 +64,20 @@ public ReviveManager(SchedulerDriver driver, TokenBucket tokenBucket) {
...
 
         if (!newCandidates.isEmpty()) {
             if (tokenBucket.tryAcquire()) {
-                logger.info("Reviving offers.");
-                driver.reviveOffers();
+                logger.info(
+                        "Reviving offers with candidates, old: {}, current: {}, new:{}",
+                        this.candidates,
+                        currCandidates,
+                        newCandidates); driver.reviveOffers();
```
[the remark this answers is not in the document]
Comment (gabrielhartmann) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/ReviveManager.java]: oops, thanks

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/ReviveManager.java [hunk 2]
```diff
@@ -83,15 +86,15 @@ public void revive(Collection<Step> steps) {
...
 
-        candidates = currCandidates;
+        this.candidates = currCandidates;
     }
 
+
     /**
      * Returns candidates which potentially need new offers.
      */
     private Set<WorkItem> getCandidates(Collection<Step> steps) {
```
Comment (benclarkwood) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/ReviveManager.java]: I'd call this `toWorkItems(Collection<Step> steps)` now that it doesn't have any conditional logic.

Comment (benclarkwood) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/ReviveManager.java]: (Or just do the conversion inline in `revive` and drop this method)

**Comment by benclarkwood** [comment]
@gabrielhartmann what you just pushed should fix those Canary strategy failures, yah?

**Comment by gabrielhartmann** [comment]
@benclarkwood: yes

**Pull request merged**